core/primitives/src/upgrade_schedule.rs

```diff
@@ -1,245 +1,458 @@
-    next_epoch_protocol_version: ProtocolVersion,
-    // Latest protocol version supported by this client.
-    client_protocol_version: ProtocolVersion,
-    // Point in time when voting for client_protocol_version version is expected
-    // to start.  Use `Default::default()` to start voting immediately.
-    voting_start: ProtocolUpgradeVotingSchedule,
-) -> ProtocolVersion {
-    if next_epoch_protocol_version >= client_protocol_version {
-        client_protocol_version
-    } else if voting_start.is_in_future() {
-        // Don't announce support for the latest protocol version yet.
-        next_epoch_protocol_version
-    } else {
-        // The time has passed, announce the latest supported protocol version.
-        client_protocol_version
+    /// This method returns the protocol version that the node should vote for.
+    pub(crate) fn get_protocol_version(
+        &self,
+        now: DateTime<Utc>,
+        // Protocol version that will be used in the next epoch.
+        next_epoch_protocol_version: ProtocolVersion,
+    ) -> ProtocolVersion {
+        if next_epoch_protocol_version >= self.client_protocol_version {
+            return self.client_protocol_version;
+        }
+
+        if self.schedule.is_empty() {
+            return self.client_protocol_version;
+        }
+
+        // The datetime values in the schedule are sorted in ascending order.
+        // Find the last datetime value that is less than the current time. The
+        // schedule is sorted and the last value is the client_protocol_version
+        // so we are guaranteed to find a correct protocol version.
+        let mut result = next_epoch_protocol_version;
+        for (time, version) in &self.schedule {
+            if now < *time {
+                break;
+            }
+            result = *version;
+        }
+
+        result
+    }
+
+    /// Returns the schedule. Should only be used for exporting metrics.
+    pub fn schedule(&self) -> &Vec<(chrono::DateTime<Utc>, ProtocolVersion)> {
+        &self.schedule
+    }
+
+    /// A helper method to parse the datetime string.
+    pub fn parse_datetime(s: &str) -> Result<DateTime<Utc>, chrono::ParseError> {
+        let datetime = NaiveDateTime::parse_from_str(s, "%Y-%m-%d %H:%M:%S")?;
+        let datetime = DateTime::<Utc>::from_naive_utc_and_offset(datetime, Utc);
+        Ok(datetime)
+    }
+
+    // Parse the protocol version override from the environment.
+    // The format is comma separate datetime:=version pairs.
+    fn parse_override(
+        override_str: &str,
+    ) -> Result<ProtocolUpgradeVotingScheduleRaw, ProtocolUpgradeVotingScheduleError> {
+        // The special value "now" means that the upgrade should happen immediately.
+        if override_str.to_lowercase() == "now" {
+            return Ok(vec![]);
+        }
+
+        let mut result = vec![];
+        let datetime_and_version_vec = override_str.split(',').collect::<Vec<_>>();
```

Comment:
nit, feel free to ignore. there's not a real need to be allocating a `Vec`, and you can just iterate over slices of the already allocated str:

```diff
diff --git a/core/primitives/src/upgrade_schedule.rs b/core/primitives/src/upgrade_schedule.rs
index a1f056026..84830e7ae 100644
--- a/core/primitives/src/upgrade_schedule.rs
+++ b/core/primitives/src/upgrade_schedule.rs
@@ -152,8 +152,7 @@ impl ProtocolUpgradeVotingSchedule {
         }
 
         let mut result = vec![];
-        let datetime_and_version_vec = override_str.split(',').collect::<Vec<_>>();
-        for datetime_and_version in datetime_and_version_vec {
+        for datetime_and_version in override_str.split(',') {
             let datetime_and_version = datetime_and_version.split('=').collect::<Vec<_>>();
             let [datetime, version] = datetime_and_version[..] else {
                 let input = format!("{:?}", datetime_and_version);
```

same is true for the `datetime_and_version.split('=').collect::<Vec<_>>();` below, where you dont really need to allocate anything to see if there's an `=` in the string and see if the left side is a valid datetime and the right side is a valid protocol version. Again, not a big deal so dont spend too much time on that though